### PR TITLE
Add transaction simulation metrics

### PR DESCRIPTION
**Pull request by matthewwalsh0**
## **Description**

Add metrics related to transaction simulation via a new `useSimulationMetrics` hook used in the `SimulationDetails` component.

This adds additional properties to the core transaction events, in addition to a new `Incomplete Asset Displayed` event.

Add bulk support to the `useName` and `useDisplayName` hooks in order to process multiple values.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/23636?quickstart=1)

## **Related issues**

Fixes: [#2192](https://github.com/MetaMask/MetaMask-planning/issues/2192)

## **Manual testing steps**

1. Login to segment site.
2. Apply associated segment key in `.metamaskrc`.
3. Create swap transaction via any swap dApp.
4. Reject or approve transaction.
5. Verify expected events in segment debug.

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/components/app/simulation-details/useSimulationMetrics.ts [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    ...getSensitiveProperties(receivingAssets, 'simulation_receiving_assets_'),
+    ...getSensitiveProperties(sendingAssets, 'simulation_sending_assets_'),
+  };
+
+  useEffect(() => {
+    updateTransactionEventFragment(
+      { properties, sensitiveProperties },
+      transactionId,
+    );
+  }, [dispatch, transactionId, JSON.stringify(properties)]);
```
Comment (OGPoyraz) [inline ui/components/app/simulation-details/useSimulationMetrics.ts:119]: Should we just remove `dispatch` here?
Also what's the reason of using `JSON.stringify` here?

Comment (matthewwalsh0) [reply ui/components/app/simulation-details/useSimulationMetrics.ts]: Good spot, `dispatch` was used in an older implementation, removed.

The `JSON.stringify` is a means to use a deep equal comparison on the parameters as a dependency to the effect, so we don't update unless the params themselves have changed. It requires primitive values hence this provides it as a string.

I could follow the entire rendering stack backwards and use `useMemo` etc but ultimately none of it is performance heavy so this is just a final guard to limit how often we update via the metrics controller.

Comment (OGPoyraz) [reply ui/components/app/simulation-details/useSimulationMetrics.ts]: One more small concern, `JSON.stringify` does not guarantee the order of properties in the serialized string. If the properties of the object are the same but their order changes, may produce a different string, causing the effect to re-run unnecessarily.

**Review by vinistevam — DISMISSED** [review]

File: ui/components/app/simulation-details/useSimulationMetrics.ts (hunk 1 above)
Comment (matthewwalsh0) [reply ui/components/app/simulation-details/useSimulationMetrics.ts]: I don't believe it guarantees any specific order, but it does provide a consistent / stable order. Meaning the same object should serialize the same way consistently which is what we need.

**Comment by sleepytanya** [comment]
Transaction simulation metrics in Segment debug:
Screenshot 2024-03-22 at 11 13 42 AM
Screenshot 2024-03-22 at 11 19 04 AM

**Review by OGPoyraz — DISMISSED** [review]

**Review by OGPoyraz — DISMISSED** [review]

**Review by vinistevam — APPROVED** [review]

**Review by OGPoyraz — APPROVED** [review]

**Review by dbrans — APPROVED** [review]
Just for good measure

**Pull request merged**